### DISCUSSION TITLE
Rails Event Store browser on Heroku not loading due to 404 Not Found errors

### DISCUSSION THREAD

**Discussion by jkwuc89** [RailsEventStore/rails_event_store#1519] [Q&A] [answered]
My team followed [these instructions](https://railseventstore.org/docs/v2/browser/) to install and configure the browser. Inside `config/routes.rb`, we use the following to mount the browser route. This is done inside the `admin` namespace.
```ruby
mount RailsEventStore::Browser => '/event_store'
```
When we run our Rails app locally, the event store browser works as expected when browsing to http://localhost:3000/admin/event_store. When we attempt to load the browser inside our app when deployed to Heroku, we get a blank page with these errors:

<img width="940" alt="Screenshot 2023-01-13 at 1 12 49 PM" src="https://user-images.githubusercontent.com/1555064/212391639-31c93f0b-67ee-4013-8166-5ff6bf16c490.png">

I went through the Heroku build log and did not see anything amiss. I'm hoping someone has run into this and knows how to fix it. Thanks in advance.

**Comment by mostlyobvious** [comment]
Hi Keith! Can you please tell what is your rails_event_store version and can you also paste bigger `routes.rb` snippet that shows how this namespace is done?

**Comment by jkwuc89** [comment]
We are currently using rails_event_store version 2.7.0. Here is a bigger `routes.rb` snippet that shows the namespace.

```ruby
namespace :admin do
    constraints ->(request) { Auth::AdminService.new(session: request.session).admin? } do
      mount Sidekiq::Web => '/sidekiq'
      mount RailsEventStore::Browser => '/event_store'
    end
end
```

`admin?` above is a method we use to determine if a user is configured as an admin in our system.

**Comment by lukaszreszke** [comment]
Hi @jkwuc89,

We tried to reproduce your issue but we weren't able to. We used new rails app with RES 2.7.0 and following `routes.rb`
```ruby
Rails.application.routes.draw do
  namespace :admin do
    constraints ->(request) { true } do
      mount RailsEventStore::Browser => '/event_store'
    end
  end
end
```
It would be much easier to investigate if provide minimal configuration app that reproduces this issue.

**Reply by robinaugh** [reply]
Hi @lukaszreszke I'm on @jkwuc89's team and I'm looking into this issue a bit more. It looks like the problem is that the event store is rendering its CSS and JS using our herokuapp.com URLs rather than the domain we're using on top of Heroku.

i.e.

```html
<script type="text/javascript" src="https://our-site.herokuapp.com/admin/event_store/ruby_event_store_browser.js"></script>
```

Should be

```html
<script type="text/javascript" src="https://our-site.com/admin/event_store/ruby_event_store_browser.js"></script>
```

It looks like providing a `path` argument to `RubyEventStore::Browser::App.for()` is [deprecated](https://github.com/RailsEventStore/rails_event_store/blob/ed055df7b8a090ddda04eb5ad4cd9d51bb1ce0ca/ruby_event_store-browser/lib/ruby_event_store/browser/app.rb#L24-L31). What's the recommended way to specify the right URL here when mounting the RES browser? Thanks!

**Reply by mostlyobvious** [reply]
The reason why `path` and `host` became deprecated is because they're irrelevant. These values can be extracted from the `Rack::Request` and this is enough to figure out host and mount point:
https://github.com/RailsEventStore/rails_event_store/blob/c876ab20b864729a03d63a6c2e059750faf8320f/ruby_event_store-browser/spec/urls_spec.rb#L24-L33

I've just checked on a Heroku-deployed application with a custom domain:

```ruby
  namespace :admin do
    authenticated :user, lambda { |u| u.admin? } do
      mount RailsEventStore::Browser, at: "/res", as: "res"
    end
  end
```

```
$ dig +short app.redacted-domain.io IN CNAME
redacted-heroku-dns-string.herokudns.com.
```

The browser is on `https://app.redacted-domain.io/admin/res` and requests:
```
https://app.redacted-domain.io/admin/res/ruby_event_store_browser.css
https://app.redacted-domain.io/admin/res/ruby_event_store_browser.js
https://app.redacted-domain.io/admin/res/bootstrap.js
```

Do you perhaps have a more complex routing setup or additional Rack middleware that could change `Rack::Request#base_url` and `Rack::Request#script_name` right before such request is passed to the browser in the application stack?

**Reply by robinaugh** [reply, marked as answer]
Closing the loop here - turns out the reason the rails event store browser was not working for us was because we were using nginx to proxy requests from a SPA javascript app back to our rails app. As soon as we retired that setup (by getting rid of the SPA 🎉 ), the rails event store browser worked just fine!
